### ISSUE TITLE
SmtpSessionManager lock

### ISSUE THREAD

**Issue by tinohager** [cosullivan/SmtpServer#228]
I think it would make sense to replace the `lock` logic with a `semaphore`.

https://github.com/cosullivan/SmtpServer/blob/cffd7df707cd0108e816b26659b277cb15a35db9/Src/SmtpServer/SmtpSessionManager.cs#L81

https://medium.com/@tyschenk20/mixing-traditional-locks-with-async-code-in-c-27431f857e01

**Comment by cosullivan** [comment]
I think this would need benchmarking before being changed as I think lock() under some cases could be faster, not to mention the code is less verbose so that is better too.

**Comment by tinohager** [comment]
One option here would also be to work with a `ConcurrentDictionary`

**lock**

| Method | Mean         | Error      | StdDev     | Gen0     | Gen1     | Allocated  |
|------- |-------------:|-----------:|-----------:|---------:|---------:|-----------:|
| Send1  |     84.76 us |   1.212 us |   1.134 us |   1.2207 |        - |   15.91 KB |
| Send2  |    444.31 us |   1.905 us |   1.591 us |   2.9297 |        - |   34.36 KB |
| Send3  |  7,255.99 us | 111.949 us |  99.239 us |   7.8125 |        - |   157.5 KB |
| Send4  | 15,390.39 us | 301.138 us | 281.685 us | 359.3750 | 296.8750 | 3778.39 KB |

**ConcurrentDictionary**

```cs
readonly ConcurrentDictionary<SmtpSessionHandle, bool> _sessions = new ConcurrentDictionary<SmtpSessionHandle, bool>();
```

| Method | Mean         | Error      | StdDev     | Gen0     | Gen1     | Allocated |
|------- |-------------:|-----------:|-----------:|---------:|---------:|----------:|
| Send1  |     85.17 us |   1.653 us |   2.525 us |   1.2207 |        - |  15.84 KB |
| Send2  |    447.85 us |   2.267 us |   1.770 us |   2.9297 |        - |  34.66 KB |
| Send3  |  7,164.46 us | 100.479 us |  89.072 us |   7.8125 |        - | 158.24 KB |
| Send4  | 15,590.24 us | 293.097 us | 259.822 us | 359.3750 | 234.3750 | 3782.4 KB |

**ConcurrentDictionary with Guid**

```cs
readonly ConcurrentDictionary<Guid, SmtpSessionHandle> _sessions = new ConcurrentDictionary<Guid, SmtpSessionHandle>();
```

| Method | Mean         | Error      | StdDev     | Gen0     | Gen1     | Allocated  |
|------- |-------------:|-----------:|-----------:|---------:|---------:|-----------:|
| Send1  |     84.59 us |   0.626 us |   0.555 us |   1.2207 |        - |   15.85 KB |
| Send2  |    459.72 us |   8.427 us |   7.470 us |   2.9297 |        - |   34.73 KB |
| Send3  |  7,134.86 us |  54.502 us |  45.512 us |   7.8125 |        - |  155.87 KB |
| Send4  | 15,134.01 us | 123.949 us | 115.942 us | 375.0000 | 281.2500 | 3788.69 KB |

**Comment by cosullivan** [comment]
So the mean performance then isn't any difference, but I guess the StdDev here probably makes the difference.

Maybe ConcurrentDictionary using a SessionId would be the best way, however, its probably better to put SessionId on the ISessionContext as the ISessionContext is passed all over so might be useful for tracking.

**Comment by tinohager** [comment]
fixed with [11.0.0-beta1](https://www.nuget.org/packages/SmtpServer/11.0.0-beta1)
